src/Datadog.Trace/Util/DbCommandCache.cs

```diff
@@ -0,0 +1,104 @@
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Data;
+using System.Data.Common;
+using System.Threading;
+using Datadog.Trace.Logging;
+using Datadog.Trace.Vendors.Serilog;
+
+namespace Datadog.Trace.Util
+{
+    internal static class DbCommandCache
+    {
+        internal const int MaxConnectionStrings = 100;
+
+        private static readonly ILogger Log = DatadogLogging.GetLogger(typeof(DbCommandCache));
+
+        private static ConcurrentDictionary<string, KeyValuePair<string, string>[]> _cache
+            = new ConcurrentDictionary<string, KeyValuePair<string, string>[]>();
+
+        /// <summary>
+        /// Gets or sets the underlying cache, to be used for unit tests
+        /// </summary>
+        internal static ConcurrentDictionary<string, KeyValuePair<string, string>[]> Cache
+        {
+            get
+            {
+                return _cache;
+            }
+
+            set
+            {
+                _cache = value;
+            }
+        }
+
+        public static KeyValuePair<string, string>[] GetTagsFromDbCommand(IDbCommand command)
+        {
+            var connectionString = command.Connection.ConnectionString;
+            var cache = _cache;
+
+            if (cache != null)
+            {
+                if (cache.TryGetValue(connectionString, out var tags))
+                {
+                    // Fast path: it's expected that most calls will end up in this branch
+                    return tags;
+                }
+
+                if (cache.Count <= MaxConnectionStrings)
+                {
+                    // Populating the cache. This path should be hit only during application warmup
+                    // ReSharper disable once ConvertClosureToMethodGroup -- Lambdas are cached by the compiler
+                    return cache.GetOrAdd(connectionString, cs => ExtractTagsFromConnectionString(cs));
```

Comment:
Just curious, is the "Resharper disable" comment just to avoid writing it like this? If so, is there a performance reason for this, or just cosmetic?
```suggestion
                    return cache.GetOrAdd(connectionString, ExtractTagsFromConnectionString);
```

Reply:
Yes, the comment is there to prevent Resharper from rewriting into a method group. The delegate is cached when using lambdas but not when using method groups, so that would cause an additional heap allocation at each call.
https://github.com/dotnet/roslyn/issues/5835

Reply:
TIL. Thanks!